### DISCUSSION TITLE
QEMU emulation package question

### DISCUSSION THREAD

**Discussion by aserper** [e-m-b-a/emba#460] [Q&A] [answered]
Hi,
I'm trying to understand but I couldn't find it in the docs:
When EMBA is trying to do a full system qemu emulation, what determines the amount of attempts that are made to generate an emulation package? Usually, EMBA generates more than one package - what is it based on?

Thanks

**Comment by m-1-k-3** [comment]
EMBA generates for every working (TCP or ICMP) emulation attempt one package. If some open port is detected EMBA will stop trying. If a port is opened then you usually would use this package. You get a final overview of all the emulation results:

![image](https://user-images.githubusercontent.com/497520/214778986-113fdc51-5f9f-420f-b6ea-e16d57debd82.png)

**Reply by aserper** [reply]
Thanks for clarifying. Is there any documentation somewhere (which isn't the code itself :wink:) to how emba tries to build the package?

**Comment by m-1-k-3** [comment, marked as answer]
There is some basic information on the emulator in the wiki [here](https://github.com/e-m-b-a/emba/wiki/System-emulation). But these details are not that detailed as you probably needed it. 

At the end the following steps are performed:
* EMBA tries to emulate the firmware in a default config to identify the network configuration
* Next step is to emulate the system multiple times with different possible configs
* On every run it tries to identify the reachability with ICMP and network services
* If the system is reachable via a network service EMBA stops and generates an archive
* If the system is reachable only via ICMP an archive is generated but EMBA tries the next config with the goal to get the system with another config further

The archive is quite simple. It is just the kernel EMBA was using, the filesystem EMBA generated and a startup wrapper script (run.sh) which configures the network and starts the emulation process. You can also just copy the steps to a terminal and it should work.

Hope that helps a bit.

**Reply by m-1-k-3** [reply]
Missed this detail. EMBA checks for more than 1 TCP service as stop criteria. If only ICMP and **one** TCP service is detected via the network EMBA will also check the other configs to improve the results.

**Reply by aserper** [reply]
Perfect, that's what I was looking for! Thank you.
